### ISSUE TITLE
Handling large number of connections

### ISSUE THREAD

**Issue by gencer** [4commerce-technologies-AG/midi-smtp-server#13]
**Describe the question**
Hi @TomFreudenberg, Is it possible to handle a large number of connections without blocking? In your examples you mentioned about 4 simultaneous connections. Let's say we want to handle large number of connections. What could be the most proper way of doing this? Manually giving a number like 1024/2048 or is there any way to make sure we maximize available connections so we did not miss any e-mail. In any way, giving 1048 for connections is acceptable? (as an example)

Basically I do not want to drop any connections when I processing already incoming and processing connections. I will use background tasks for I/O tasks, however, I would like to make sure about that too.

**Expected behavior**
Currently a fixed number of connections is allowed.

**System (please complete the following information):**
 - OS: Linux*Debian 9
 - Ruby: 2.5.1 MRI
 - MidiSmtpServer version Latest

**Additional information**
-No additional info-

**Comment by TomFreudenberg** [comment]
Hi @gencer,

thanks for posting and using this library. I will try to answer your question:

<br>

> Is it possible to handle a large number of connections without blocking?

Not sure what you mean with "blocking" but max-connections does not mean to block the IP connection, it does mean "simultaniously" processing communication.

look at: https://github.com/4commerce-technologies-AG/midi-smtp-server/blob/master/lib/midi-smtp-server.rb#L313-L317

If there are more IP connections (threads) than simultaniously processing channels, those threads will be accepted (established tcp connect from client to server) but not processed by SMTP communication.

You can test this: Just allow (1) connection and connect two times by telnet and enter "HELO test" on both. The first telnet is processed while the second will wait until timeout or first ist finished (QUIT).

So with that you won't loose any incoming connection whatever _max_connection_ is set to.

<br>

> In any way, giving 1048 for connections is acceptable?

It just depends on your system. If RAM and CPU is fine you can run as much threads simultaniously as you like. From my point of view it depends on how many messages you have to proceed per time period.

Lets try to figure out for a huge mail server 1.000.000 mails per day. Divided by seconds per day (24 * 60 * 60 = 86.400) this are 11.5 mails per second. Lets define the average processing time per mail 15 seconds (it is long time). then you might have an overlap for 15 times 11.5 connections simultaniously. If that is expected, than MAX_CONNECTIONS of 172 should fit.

If you need 1.000.000 mail per hour than propably 416 simultaniously processed threads should be fine.

In any case, no connection is dropped - just not processed until free queue.

<br>

> Basically I do not want to drop any connections when I processing already incoming and processing connections. I will use background tasks for I/O tasks, however, I would like to make sure about that too.

Not sure what you are looking for, but maybe the above information already answer that too.

<br>

Do not hesitate to ask further questions.

Cheers
Tom

**Comment by gencer** [comment]
@TomFreudenberg, This was exactly what I was referring to. Thank you so much for your detailed response!

After seeing that part of #L313-L317, It makes sense.

P.S.: Thank you for this great library! I did similar job in Go but due to some special "needs", I need to adapt this on Ruby. I found your implementations much more flexible and easy for adapt and developing.

**Comment by TomFreudenberg** [comment]
Hi @gencer 

I close that issue now.

Good luck for your project
Tom

**Comment by TomFreudenberg** [comment]
Hi @gencer 

I have add a new commit tonight https://github.com/4commerce-technologies-AG/midi-smtp-server/commit/6756517fefd31941585b88fd4cfca7c65fc73f9b

This will change what I have written here, because now max_connections does handle the number of max TCP connections.

So you should now check for max_connections and max_processings.

If you only set max_connections, then max_processings will be set equal.

Cheers
Tom

**Comment by TomFreudenberg** [comment]
Hi @gencer 

I made another commit this morning (https://github.com/4commerce-technologies-AG/midi-smtp-server/commit/538d7d89347b02ad47d96de4f4de318b3795311d) and changed the order of max_processings and max_connections to their intents. So by now if you leave your sources untouched everything works as expected in previous releases as well. But with 2.3.0 you also may limit the number of concurrent TCP connections. That was not possible to set before.

Cheers
Tom

**Comment by gencer** [comment]
Hi @TomFreudenberg, I hope you are having a great day!

This changes works perfectly. I am not going to use max_processings for now on my primary server due to I do not want to drop any connection but hang them so they will queued. However, one of my server has a limited resources which is max_processings will work beautifully.

**Comment by gencer** [comment]
Whoa, Why I didnt read that sentence:

> That error code will normally ensure, that the sender would try again after a while.

Which is acceptable.

**Comment by TomFreudenberg** [comment]
Hi @gencer 

just to make you sure:

The only value you are currently interested in is `max_processings`.

Per default `max_connections` is set to NIL what means and allows an unlimited number of  concurrent TCP connections (threads) to be established.

The value of `max_processings` will handle the queueing, so that sessions more than `max_processings` will have to wait until slot is available.

If you also want to limit the concurrent TCP connections (threads) then you may specifiy also `max_connections`.

In neither way you _should_ loose any mails but at least it depends on that all systems between will work conform to the mail RFCs.

Hope that makes it clear view to your options.

Cheers
Tom
